firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java

```diff
@@ -0,0 +1,168 @@
+    }
+
+    if (bitmap.length > 0) {
+      // Only empty bloom filter can have 0 hash count.
+      if (hashCount <= 0) {
+        throw new IllegalArgumentException("Invalid hash count: " + hashCount);
+      }
+    } else {
+      if (hashCount < 0) {
+        throw new IllegalArgumentException("Invalid hash count: " + hashCount);
+      }
+
+      // Empty bloom filter should have 0 padding.
+      if (padding != 0) {
+        throw new IllegalArgumentException("Invalid padding when bitmap length is 0: " + padding);
+      }
+    }
+    this.bitmap = bitmap;
+    this.hashCount = hashCount;
+    this.bitCount = bitmap.length * 8 - padding;
+    this.md5HashMessageDigest = createMd5HashMessageDigest();
+  }
+
+  @VisibleForTesting
+  int getBitCount() {
+    return this.bitCount;
+  }
+
+  /**
+   * Check whether the given string is a possible member of the bloom filter. It might return false
+   * positive result, ie, the given string is not a member of the bloom filter, but the method
+   * returned true.
+   *
+   * @param value the string to be tested for membership.
+   * @return true if the given string might be contained in the bloom filter.
+   */
+  public boolean mightContain(@NonNull String value) {
+    // Empty bitmap or empty value should always return false on membership check.
+    if (this.bitCount == 0  || value.isEmpty()) {
+      return false;
+    }
+
+    byte[] hashedValue = md5HashDigest(value);
+    if (hashedValue.length != 16) {
+      throw new RuntimeException(
+          "Invalid md5 hash array length: " + hashedValue.length + " (expected 16)");
+    }
+
+    long hash1 = getLongLittleEndian(hashedValue, 0);
+    long hash2 = getLongLittleEndian(hashedValue, 8);
+
+    for (int i = 0; i < this.hashCount; i++) {
+      int index = this.getBitIndex(hash1, hash2, i);
+      if (!this.isBitSet(index)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /** Hash a string using md5 hashing algorithm, and return an array of 16 bytes. */
+  @NonNull
+  private byte[] md5HashDigest(@NonNull String value) {
+    return md5HashMessageDigest.digest(value.getBytes(StandardCharsets.UTF_8));
+  }
+
+  @NonNull
+  private MessageDigest createMd5HashMessageDigest() {
+    try {
+      return  MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Missing MD5 MessageDigest provider.", e);
+    }
+  }
+
+  /** Interpret 8 bytes into a long, using little endian 2’s complement. */
+  private static long getLongLittleEndian(@NonNull byte[] bytes, int offset) {
+    long result = 0;
+    for (int i = 0; i < 8; i++) {
+      result |= (bytes[offset + i] & 0xFFL) << (i * 8);
+    }
+    return result;
+  }
+
+  /**
+   * Calculate the ith hash value based on the hashed 64 bit unsigned integers, and calculate its
+   * corresponding bit index in the bitmap to be checked.
+   */
+  private int getBitIndex(long hash1, long hash2, int index) {
```

Comment:
Consider renaming the `index` argument to something like `hashIndex` or `hashNum`. There are two "indexes" being discussed in the documentation and IMO it's helpful to be clear which index is being referred to.